pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,496 @@
+		// $ ovs-vsctl add-br br0
+		// $ #For ERSPAN type 2 (version I)
+		// $ ovs-vsctl add-port br0 at_erspan0 -- \
+		//        set int at_erspan0 type=erspan options:key=1 \
+		//        options:remote_ip=172.31.1.1 \
+		//        options:erspan_ver=1 options:erspan_idx=1
+
+		// $ #For ERSPAN type 3 (version II)
+		// $ ovs-vsctl add-port br0 at_erspan0 -- \
+		//        set int at_erspan0 type=erspan options:key=1 \
+		//        options:remote_ip=172.31.1.1 \
+		//        options:erspan_ver=2 options:erspan_dir=1 \
+		//        options:erspan_hwid=4
+
+		config := device.ERSPANConfig
+		tunnelID := config.TunnelID
+		remoteIP := config.RemoteIP
+		version := config.Version
+		index := config.Index
+		dir := config.Dir
+		hardwareID := config.HardwareID
+
+		externalIDs := map[string]interface{}{
+			interfacestore.AntreaInterfaceTypeKey: interfacestore.AntreaUnset,
+			"erspan_ver":                          version,
+			"key":                                 1,
+		}
+		if version == 1 {
+			if index != nil {
+				externalIDs["erspan_idx"] = index
+			}
+		} else if version == 2 {
+			externalIDs["erspan_dir"] = dir
+			externalIDs["erspan_hwid"] = hardwareID
+		}
+		portUUID, err := c.ovsBridgeClient.CreateTunnelPortExt(device.Name,
+			ovsconfig.ERSpanTunnel, 0, false, "", remoteIP, "", externalIDs)
+		if err != nil {
+			return 0, err
+		}
+		ofPortTmp, err = c.ovsBridgeClient.GetOFPort(device.Name, false)
+		if err != nil {
+			return 0, err
+		}
+		itf := interfacestore.NewTunnelInterface(device.Name, ovsconfig.ERSpanTunnel, nil, false)
+		itf.OVSPortConfig = &interfacestore.OVSPortConfig{PortUUID: portUUID, OFPort: ofPortTmp}
+		c.interfaceStore.AddInterface(itf)
+		ofPort = uint32(ofPortTmp)
+
+		klog.InfoS("Create ERSPAN", "tunnelID", tunnelID, "remoteIP", remoteIP, "version", version, "index", index,
+			"dir", dir, "hardwareID", hardwareID)
+	}
+	return
+}
+
+func (c *Controller) listOVSPorts() error {
+	ovsPorts, err := c.ovsBridgeClient.GetPortList()
+	if err != nil {
+		return err
+	}
+	ports := make(map[string]ovsconfig.OVSPortData, len(ovsPorts))
+	for i := range ovsPorts {
+		ports[ovsPorts[i].Name] = ovsPorts[i]
+	}
+	c.ovsPorts = ports
+	return nil
+}
+
+func (c *Controller) syncTrafficControl(tcName string) (err error) {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing TrafficControl", "TrafficControl", tcName, "durationTime", time.Since(startTime))
+	}()
+
+	tc, err := c.trafficControlLister.Get(tcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			eState, exist := c.getTcState(tcName)
+			if !exist {
+				return nil
+			}
+			if err := c.uninstallTrafficControl(tcName, eState); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	tcState, exist := c.getTcState(tcName)
+
+	if !exist {
+		tcState = c.createTcState(tcName)
+		var dstInPort, dstOutPort uint32
+		if err := c.listOVSPorts(); err != nil {
```

Comment:
1. It's not thread-safe to update `ovsPorts` concurrently
2. It's not worth to list all ports everytime it syncs a TC object, why don't query a port when it's unknown?

Reply:
Updated.